decoders/sent/pd.py

```diff
@@ -0,0 +1,805 @@
+
+        if self.pulse_len_in_us + self.clock_time_time < min_len:
+            return field_name + ": pulse too short"
+
+        if self.pulse_len_in_us - self.clock_time_time > max_len:
+            return field_name + ": pulse too long"
+
+        return field_name + ": unknown"
+
+    def explain_bad_nibble(self):
+        return self._explain('Nibble', self.nibble_min_len_us, self.nibble_max_len_us)
+
+    def explain_bad_calibration(self):
+        return self._explain('Calibration', self.calibration_pulse_len_us, self.calibration_pulse_len_us)
+
+    def explain_bad_pause(self):
+        return self._explain('Pulse Pause', self.nibble_min_len_us, self.pause_pulse_max_len_us)
+
+    def put_error(self, string):
+        self.put_error_at(self.pulse_falling0, self.pulse_falling1, string)
+
+    def put_error_at(self, start, end, string):
+        self.put_field_at(start, end, Decoder.ERROR_ANN_INDEX, [string])
+
+    def put_field(self, field_index, strings):
+        self.put_field_at(self.pulse_falling0, self.pulse_falling1, field_index, strings)
+
+    def put_field_at(self, start, end, field_index, strings):
+        self.put(start, end, self.output_ann, [field_index, strings])
+
+    def _set_slow_or_fail(self):
+        format = self.options['slow-format']
+
+        if Sent.SLOW_CHANNEL_FORMAT_NONE == format:
+            self.slow_protocol = SlowProtocolNone()
+        elif Sent.SLOW_CHANNEL_FORMAT_SHORT == format:
+            self.slow_protocol = SlowProtocolShort()
+        elif Sent.SLOW_CHANNEL_FORMAT_ENHANCED_C0 == format:
+            self.slow_protocol = SlowProtocolEnhanced(0)
+        elif Sent.SLOW_CHANNEL_FORMAT_ENHANCED_C1 == format:
+            self.slow_protocol = SlowProtocolEnhanced(1)
+        else:
+            raise SlowFormatError(f'unknown format {format}')
+
+        self.slow_protocol.decoder = self
+
+    @staticmethod
+    def format_channel_value(no, value, hex_chars):
+        dec_value = str(value)
+        hex_value = "0x{0:0{1}X}".format(value, hex_chars)
+
+        return [f'Channel {no}: {dec_value} ({hex_value})', f'Ch{no}: {dec_value} ({hex_value})', f'Ch{no}: ({hex_value})']
+
+    def decode(self):
+        if not self.samplerate:
+            raise SamplerateError('Cannot decode without samplerate.')
+
+        self.sample_rate_to_us_conversion_factor = 1_000_000 / self.samplerate
+
+        self.clock_time_time = self.options['clock-tick-time']
+        fast_format = self.options['fast-format']
+        pulse_pause_mode = self.options['pulse-pause']
+
+        if self.clock_time_time < 1:
+            raise ClockTickTimeError('Clock Tick Time must be greater or equal to 1 microsecond')
+
+
+        self._set_slow_or_fail()
+
+        nibbles = Sent.FAST_FORMAT_TO_FRAME_LENGTH[fast_format]
+
+
+        self.clock_time_time_half = self.clock_time_time / 2
+        self.nibble_min_len_us = 12 * self.clock_time_time
+        self.nibble_max_len_us = 27 * self.clock_time_time
```

Comment:
are these magic constants (12, 27, etc) derived or directly from the spec ? In any case I think they should probably  be explained a bit more

Reply:
Converted to constants in Sent class, added spec references